Add init_with method to Mail::Body

lesleh opened the pull request:
Fixes the case where a `PartsList` gets serialized as an `Array`, but does not get correctly deserialized back to a `PartsList` instance. The `init_from` would ordinarily be added to the `PartsList` class but this does not work due to the use of `SerializeClass(Array)` so instead I assign it on deserialization of `Mail::Body`.

CC @jeremy

Related to https://github.com/mikel/mail/issues/1038

jeremy commented on lib/mail/body.rb:
```diff
@@ -50,6 +50,11 @@ def initialize(string = '')
     end
 
+    def init_with(coder)
+      coder.map.each { |k, v| instance_variable_set(:"@#{k}", v) }
```
Can we call `super` to pull in this implementation, or do we need to implement it ourselves?

lesleh replied:
There's no implementation of it in the super class to call so I left it out.

jeremy commented on lib/mail/body.rb:
```diff
@@ -50,6 +50,11 @@ def initialize(string = '')
 
+    def init_with(coder)
+      coder.map.each { |k, v| instance_variable_set(:"@#{k}", v) }
+      @parts = Mail::PartsList.new(coder['parts'])
```
Nice fix; retains backward compatibility.

lesleh commented:
@jeremy I added the test cases from that PR, and added one to ensure that `Mail::Message#parts` remains a `Mail::PartList` through serialisation and deserialisation.

jeremy commented:
Backported to 2-7-stable @ 3204c0b4733166b9664a552006286227dea09953
